wagtail/wagtailadmin/static_src/wagtailadmin/js/page-editor.js

```diff
@@ -361,80 +361,67 @@ $(function() {
-                            previewDoc.open();
-                            previewDoc.write(data);
-                            previewDoc.close();
-                        }
-
-                    } else {
-                        previewWindow.close();
-                        disableDirtyFormCheck();
-                        document.open();
-                        document.write(data);
-                        document.close();
-                    }
-                },
-
-                error: function(xhr, textStatus, errorThrown) {
-                    /* If an error occurs, display it in the preview window so that
-                    we aren't just showing the spinner forever. We preserve the original
-                    error output rather than giving a 'friendly' error message so that
-                    developers can debug template errors. (On a production site, we'd
-                    typically be serving a friendly custom 500 page anyhow.) */
-
-                    previewDoc.open();
-                    previewDoc.write(xhr.responseText);
-                    previewDoc.close();
-                }
-            });
+    //
+    // Preview
+    //
+    // In order to make the preview truly reliable, the preview page needs
+    // to be perfectly independent from the edit page,
+    // from the browser perspective. To pass data from the edit page
+    // to the preview page, we send the form after each change
+    // and save it inside the user session.
+
+    var $previewButton = $('.action-preview'), $form = $('#page-edit-form');
+    var previewUrl = $previewButton.data('action');
+
+    function setPreviewData() {
+        return $.ajax({
+            url: previewUrl,
+            method: 'POST',
+            data: new FormData($form[0]),
+            processData: false,
+            contentType: false
+        });
+    }
 
-        }
+    if ($previewButton.data('auto-update')) {
+        var setPreviewDataTimeout = -1;
+
+        // Form data is changed when field values are changed (change event),
+        // when HTML elements are added, modified, moved, and deleted
+        // (DOMSubtreeModified event), and we need to delay setPreviewData
+        // when typing to avoid useless extra AJAX requests
+        // (so we postpone setPreviewData when keyup occurs).
+        // TODO: Replace DOMSubtreeModified with a MutationObserver.
+        $form.on('change keyup DOMSubtreeModified', function () {
+            clearTimeout(setPreviewDataTimeout);
+            setPreviewDataTimeout = setTimeout(setPreviewData, 1000);
+        }).change();
+    }
 
+    $previewButton.click(function(e) {
+        e.preventDefault();
+        var $this = $(this);
+        var $icon = $this.filter('.icon'),
+            thisPreviewUrl = $this.data('action');
+        $icon.addClass('icon-spinner').removeClass('icon-view');
+        var previewWindow = window.open('', thisPreviewUrl);
+        previewWindow.focus();
```

Comment:
I guess the least confusing is then to follow the browser's behavior instead of explicitly calling `.focus()`? Would like to get others' opinion on this, as I know it's pretty controversial.